python/tests/test_schema_validator.py

```diff
@@ -0,0 +1,211 @@
+                feature_group_data, pandas_df, feature_group_data.features
+            )
+
+    def test_validate_schema_primary_key_null(self, pandas_df, feature_group_data):
+        pandas_df.loc[0, "primary_key"] = None
+        with pytest.raises(
+            ValueError, match="Primary key column primary_key contains null values"
+        ):
+            DataFrameValidator().validate_schema(
+                feature_group_data, pandas_df, feature_group_data.features
+            )
+
+    def test_validate_schema_string_length_exceeded(
+        self, pandas_df, feature_group_created, mocker
+    ):
+        pandas_df.loc[0, "string_col"] = "a" * 101
+        with pytest.raises(ValueError, match="String length exceeded"):
+            DataFrameValidator().validate_schema(
+                feature_group_created, pandas_df, feature_group_created.features
+            )
+
+    def test_validate_schema_feature_group_created(
+        self, pandas_df, feature_group_created, mocker
+    ):
+        pandas_df.loc[0, "string_col"] = "a" * 101
+        with pytest.raises(ValueError, match="String length exceeded"):
+            DataFrameValidator().validate_schema(
+                feature_group_created, pandas_df, feature_group_created.features
+            )
+
+    def test_validate_schema_feature_group_not_created(
+        self, pandas_df, feature_group_data
+    ):
+        # test with non existing feature group with no explicit features
+        # arrange
+        pandas_df.loc[0, "string_col"] = "a" * 101
+        initial_features = [
+            Feature("primary_key", "int"),
+            Feature("event_time", "string"),
+            Feature("string_col", "string"),
+        ]
+        feature_group_data.features = []
+        df_features = DataFrameValidator().validate_schema(
+            feature_group_data, pandas_df, initial_features
+        )
+        assert df_features[2].online_type == "varchar(101)"
+
+    def test_validate_schema_feature_group_with_features_not_created(
+        self, pandas_df, feature_group_data
+    ):
+        # test with feature group with explicit features
+        df_features = DataFrameValidator().validate_schema(
+            feature_group_data, pandas_df, feature_group_data.features
+        )
+        # assert that the online type of the string_col feature is same as explcitly set in the feature group
+        assert df_features[2].online_type == "varchar(200)"
+
+    def test_pk_null_string_length_exceeded(self, pandas_df, feature_group_data):
+        pandas_df.loc[0, "primary_key"] = None
+        pandas_df.loc[0, "string_col"] = "a" * 101
+        with pytest.raises(
+            ValueError, match="One or more schema validation errors found"
+        ):
+            DataFrameValidator().validate_schema(
+                feature_group_data, pandas_df, feature_group_data.features
+            )
+
+    def test_offline_fg(self, pandas_df, feature_group_data, caplog):
```

Comment:
What is this test doing? Offline FG should never be validated by the validator, no?

Reply:
it  was for testing that it skips validation if offline and returns the features unchanged. Also checking that warning is raised.